elytron/src/main/java/org/wildfly/extension/elytron/SSLDefinitions.java

```diff
@@ -1511,6 +1532,56 @@ protected void installedForResource(ServiceController<SSLContext> serviceControl
     }
 
+    private static final AttributeDefinition[] DYNAMIC_SSL_CONTEXT_ATTRIBUTES = new AttributeDefinition[]{AUTHENTICATION_CONTEXT_ATTRIBUTE};
+
+    static ResourceDefinition getDynamicClientSSLContextDefinition(boolean serverOrHostController) {
+        return createSSLContextDefinition(ElytronDescriptionConstants.DYNAMIC_CLIENT_SSL_CONTEXT, false, new DynamicSSLContextAddHandler(), DYNAMIC_SSL_CONTEXT_ATTRIBUTES, serverOrHostController, Stability.COMMUNITY);
+    }
+
+    private static class DynamicSSLContextAddHandler extends BaseAddHandler {
+        private DynamicSSLContextAddHandler() {
+            super(SSL_CONTEXT_RUNTIME_CAPABILITY, DYNAMIC_SSL_CONTEXT_ATTRIBUTES);
+        }
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            ServiceBuilder<SSLContext> serviceBuilder = installService(context, model);
+            ServiceName authenticationContextServiceName =
+                    AUTHENTICATION_CONTEXT_RUNTIME_CAPABILITY.getCapabilityServiceName(AUTHENTICATION_CONTEXT_ATTRIBUTE.resolveModelAttribute(context, model).asString());
+            serviceBuilder.requires(authenticationContextServiceName);
+            commonRequirements(serviceBuilder).setInitialMode(ServiceController.Mode.ACTIVE).install();
+        }
+
+        ServiceBuilder<SSLContext> installService(OperationContext context, ModelNode model) throws OperationFailedException {
+            ServiceName dynamicSslContextServiceName = SSL_CONTEXT_RUNTIME_CAPABILITY.getCapabilityServiceName(context.getCurrentAddressValue());
+            return context.getServiceTarget().addService(dynamicSslContextServiceName, new TrivialService<>(getValueSupplier(dynamicSslContextServiceName, context, model)));
+        }
+
+        protected ValueSupplier<SSLContext> getValueSupplier(ServiceName dynamicSSLContextName, OperationContext context, ModelNode model) throws OperationFailedException {
+            String authenticationContextName = AUTHENTICATION_CONTEXT_ATTRIBUTE.resolveModelAttribute(context, model).asString();
+            ServiceRegistry serviceRegistry = context.getServiceRegistry(true);
+            return () -> {
+                ServiceController<?> dynamicSSLContextService = serviceRegistry.getRequiredService(dynamicSSLContextName);
+                State dynamicSSLContextServiceState = dynamicSSLContextService.getState();
+                if (!dynamicSSLContextServiceState.equals(State.UP)) {
+                    dynamicSSLContextService.setMode(ServiceController.Mode.ACTIVE);
+                }
+                try {
+                    return new DynamicSSLContext(new DynamicSSLContextImpl(getAuthenticationContextService(serviceRegistry, authenticationContextName).getValue()));
+                } catch (GeneralSecurityException | DynamicSSLContextException e) {
+                    throw new StartException(e);
+                }
```

Comment:
This manner of obtaining the value provided by a service is not thread safe.

Reply:
@pferraro Can you please elaborate on what is not thread safe in this method? I think we obtain a value from a service like this on other places also and I am not sure what can cause issues. What should be a correct way to do this?

Reply:
> Can you please elaborate on what is not thread safe in this method? 

* Testing that the state of the ServiceController is UP does not guarantee that the service is still UP when you call Service.getValue().
* ServiceController.setMode(...) is not a blocking operation, i.e. there are no guarantees that the service will be UP when this call returns.
* Service.getValue() was deprecated several years ago to discourage use for this reason.

> I think we obtain a value from a service like this on other places

If that is the case, those other places are problematic too.

Reply:
@pferraro Thank you that makes sense. I have changed the code now and it is no longer thread unsafe. 

I have addressed all feedback now, so this can be reviewed. Thank you!